include/boost/math/fft/real_algorithms.hpp

```diff
@@ -346,6 +346,147 @@
       real_inverse_dft_prime_bruteForce_outofplace(in_first,in_last,out,sign);
   }
   
+  template <class T, class allocator_t>
+  void real_dft_composite_outofplace(
+            const T *in_first, 
+            const T *in_last, 
+            T* out, 
+            int sign,
+            const allocator_t& alloc)
+  {
+    /*
+      Cooley-Tukey mapping, intrinsically out-of-place, Decimation in Time
+      composite sizes.
+    */
+    using allocator_type = allocator_t;
+    using ComplexType = ::boost::multiprecision::complex<T>; 
+    using ComplexAllocator = typename std::allocator_traits<allocator_type>::rebind_alloc<ComplexType>;
+    
+    const long n = static_cast<long>(std::distance(in_first,in_last));
+    if(n <=0 )
+      return;
+    
+    if (n == 1)
+    {
+        out[0]=in_first[0];
+        return;
+    }
+    std::array<int,32> prime_factors;
+    const int nfactors = prime_factorization(n,prime_factors.begin());
+    
+    // reorder input
+    for (long i = 0; i < n; ++i)
+    {
+        long j = 0, k = i;
+        for (int ip=0;ip<nfactors;++ip)
+        {
+            int p = prime_factors[ip];
+            j = j * p + k % p;
+            k /= p;
+        }
+        out[j] = in_first[i];
+    }
+    
+    std::reverse(prime_factors.begin(), prime_factors.begin()+nfactors);
+    
+    //auto show = [&] (const T* beg, const T* end)
+    //{
+    //  for(;beg!=end;++beg)
+    //  {
+    //    std::cout << *beg << ", ";
+    //  }
+    //  std::cout << "\n";
+    //};
+    
+    // butterfly pattern
+    long len = 1;
+    for (int ip=0;ip<nfactors;++ip)
+    {
+      int p = prime_factors[ip];
+      long len_old = len;
+      len *= p;
+      
+      std::vector<ComplexType,ComplexAllocator> tmp(p,alloc);
+      //std::cout << "pass " << ip << "\n";
+      for (long i = 0; i < n; i += len)
+      {
+        //std::cout << "    i = " << i << "\n";
+        for(long k=0;2*k<=len_old;++k)
```

Comment:
Here lies the optimization of real-to-halfcomplex. In the complex-to-complex routine this for loop would be `for(long k=0;k<len_old;++k)`, we use instead half of the values of `k`.